prow/e2e-gcloud-build-image.sh

```diff
@@ -0,0 +1,94 @@
+SERVICE_NAME="hellogrpc.endpoints.cloudesf-testing.cloud.goog"
+CONFIG_ID="2018-11-11r0"
+
+function error_exit() {
+  # ${BASH_SOURCE[1]} is the file name of the caller.
+  echo "${BASH_SOURCE[1]}: line ${BASH_LINENO[0]}: ${1:-Unknown Error.} (exit ${2:-1})" 1>&2
+  exit ${2:-1}
+}
+
+function formImageName() {
+  local expected_version=$1
+  echo "gcr.io/${PROJECT_NAME}/endpoints-runtime-serverless:${expected_version}-${SERVICE_NAME}-${CONFIG_ID}"
+}
+
+function cleanupOldImage() {
+  local image_name=$1
+  echo "Cleaning up old image if it exists (ignore any errors in the output here)."
+  if gcloud container images describe "${image_name}"; then
+    gcloud container images delete "${image_name}"
+  fi
+}
+
+function expectImage() {
+  local image_name=$1
+  gcloud container images describe "${image_name}" || error_exit "Failed to find image: ${image_name}"
+  echo "Successfully verified image exists: ${image_name}"
+}
+
+echo "=== Test 1: Specify a fully qualified version. ==="
+EXPECTED_IMAGE_NAME=$(formImageName "2.7.0")
+cleanupOldImage "${EXPECTED_IMAGE_NAME}"
+${ROOT}/docker/serverless/gcloud_build_image \
+    -s "${SERVICE_NAME}" \
+    -c "${CONFIG_ID}" \
+    -p "${PROJECT_NAME}" \
+    -v "2.7.0"
+expectImage "${EXPECTED_IMAGE_NAME}"
+
+echo "=== Test 2: Specify a minor version. ==="
+EXPECTED_IMAGE_NAME=$(formImageName "2.4.0")
+cleanupOldImage "${EXPECTED_IMAGE_NAME}"
+${ROOT}/docker/serverless/gcloud_build_image \
+    -s "${SERVICE_NAME}" \
+    -c "${CONFIG_ID}" \
+    -p "${PROJECT_NAME}" \
+    -v "2.4"
+expectImage "${EXPECTED_IMAGE_NAME}"
+
+echo "=== Test 3: Sepcify an invalid version fails. ==="
+if ${ROOT}/docker/serverless/gcloud_build_image \
+    -s "${SERVICE_NAME}" \
+    -c "${CONFIG_ID}" \
+    -p "${PROJECT_NAME}" \
+    -v "2.11.47"; then
+  error_exit "Script should fail for invalid version."
+else
+  echo "Script failed as expected."
+fi
+
+echo "=== Test 4: Specify a custom image. ==="
+EXPECTED_IMAGE_NAME=$(formImageName "custom")
+cleanupOldImage "${EXPECTED_IMAGE_NAME}"
+${ROOT}/docker/serverless/gcloud_build_image \
+    -s "${SERVICE_NAME}" \
+    -c "${CONFIG_ID}" \
+    -p "${PROJECT_NAME}" \
+    -i "gcr.io/cloudesf-testing/apiproxy-serverless:gcloud-build-image-test"
+expectImage "${EXPECTED_IMAGE_NAME}"
+
+echo "=== Test 5: When no ESP version is specified, the script uses the latest ESPv2 release. ==="
+# Knowing the latest ESP version number is hard, it depends on what is tagged in GCR.
+# This is a chicken and egg problem, because `gcloud_build_image` uses that.
+# That means we don't have a reliable way of checking if the output is correct.
+# So just test the script passes, and allow the developer to manually verify the output.
+${ROOT}/docker/serverless/gcloud_build_image \
+    -s "${SERVICE_NAME}" \
+    -c "${CONFIG_ID}" \
+    -p "${PROJECT_NAME}"
+echo ">>> WARNING: For the test above, manually verify the output version of the image is expected."
```

Comment:
can we get the latest version from file "VERSION" that we updated for each release?

Reply:
I considered this before making the PR, but there's a few complications. I don't think it's worth trying to work around these:

1) VERSION file has the version for the next release. Current stable version is 1 below the content in the file.
2) During a release, we make a PR to increment the version before we make the release stable. In this time period, the VERSION file would be 2 ahead the current stable.
3) If we need to roll back a release (as we've done in the past), the VERSION file will be 2 (or more) ahead the current stable version.
4) If we ever release with a patch version, then it gets more complicated. Complications 1-3 are only considering minor versions, not patch.

Reply:
I see.
How about defining the office_release_version here in this script and  update release doc to add a step to update this version for each release and each rollback.

Reply:
We can, but I'm still worried it will complicate the release process. 

The release engineer needs to submit the PR **as soon as** they update the docker tag in GCR. Otherwise, the versions will be out of sync, and the test will start failing.

Honestly, I do not think it's worth making the release process more complicated for this test. But it's up to you, what do you think?

Reply:
I am concerning about if some change is changed, the script is not getting correct release version,  how do we catch this?
For now, it depends on manually run this script,  of check this e2e test log if it is passed.   Such manual step will not catch bugs.

Reply:
Discussed offline, decided this would complicate the release process too much.